pkg/retry/lib/retry.dart

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:math' as math;
+
+final _rand = math.Random();
+
+/// Object holding options for retrying a function.
+///
+/// With the default configuration functions will be retried up-to 7 times
+/// (8 attempts in total), sleeping 1st, 2nd, 3rd, ..., 7th attempt:
+///  1. 400 ms +/- 25%
+///  2. 800 ms +/- 25%
+///  3. 1600 ms +/- 25%
+///  4. 3200 ms +/- 25%
+///  5. 6400 ms +/- 25%
+///  6. 12800 ms +/- 25%
+///  7. 25600 ms +/- 25%
+///
+/// **Example**
+/// ```dart
+/// final r = RetryOptions();
+/// final response = await r.retry(
+///   // Make a GET request
+///   () => http.get('https://google.com').timeout(Duration(seconds: 5)),
+///   // Retry on SocketException or TimeoutException
+///   retryIf: (e) => e is SocketException || e is TimeoutException,
+/// );
+/// print(response.body);
+/// ```
+class RetryOptions {
+  /// Delay factor to double after every attempt.
+  ///
+  /// Defaults to 200 ms, which results in the following delays:
+  ///
+  ///  1. 400 ms
+  ///  2. 800 ms
+  ///  3. 1600 ms
+  ///  4. 3200 ms
+  ///  5. 6400 ms
+  ///  6. 12800 ms
+  ///  7. 25600 ms
+  ///
+  /// Before application of [randomizationFactor].
+  final Duration delayFactor;
+
+  /// Percentage the delay should be randomized, given as fraction between
+  /// 0 and 1.
+  ///
+  /// If [randomizationFactor] is `0.25` (default) this indicates 25 % of the
+  /// delay should be increased or decreased by 25 %.
+  final double randomizationFactor;
+
+  /// Maximum delay between retries, defaults to 30 seconds.
+  final Duration maxDelay;
+
+  /// Maximum number of attempts before giving up, defaults to 8.
+  final int attempts;
+
+  /// Create a set of [RetryOptions].
+  ///
+  /// Defaults to 8 attempts, sleeping as following after 1st, 2nd, 3rd, ...,
+  /// 7th attempt:
+  ///  1. 400 ms +/- 25%
+  ///  2. 800 ms +/- 25%
+  ///  3. 1600 ms +/- 25%
+  ///  4. 3200 ms +/- 25%
+  ///  5. 6400 ms +/- 25%
+  ///  6. 12800 ms +/- 25%
+  ///  7. 25600 ms +/- 25%
+  const RetryOptions({
+    this.delayFactor = const Duration(milliseconds: 200),
+    this.randomizationFactor = 0.25,
+    this.maxDelay = const Duration(seconds: 30),
+    this.attempts = 8,
+  });
+
+  /// Delay after [attempt] number of attempts.
+  ///
+  /// This is computed as `pow(2, attempt) * delayFactor`, then is multiplied by
+  /// between `-randomizationFactor` and `randomizationFactor` at random.
+  Duration delay(int attempt) {
+    assert(attempt >= 0, 'attempt cannot be negative');
+    if (attempt <= 0) {
+      return Duration();
+    }
+    final rf = (randomizationFactor * (_rand.nextDouble() * 2 - 1) + 1);
+    var delay = math.min(
```

Comment:
I think something like this would be more compact:
```
final delay = (math.pow(2, attempt) * delayFactor* rf);
return delay > maxDelay ? maxDelay : delay;
```

Reply:
Oh, I had a hard time figuring out of these operators was implemented...